guides/rag-chatbot-offline-evaluation.mdx

```diff
@@ -24,139 +24,237 @@ literalai>=0.0.501
+
+<CodeGroup>
+```python Run tool calls
 
-# Embedding the user query using a model from OpenAI
 @cl.step(name="Embed", type="embedding")
-async def embed(query, model="text-embedding-ada-002"):
-    embedding = await openai_client.embeddings.create(input=query, model=model)
+async def embed(question, model="text-embedding-ada-002"):
+    """
+    Embed a question using the specified model and return the embedding.
+    """
+    embedding = await openai_client.embeddings.create(input=question, model=model)
 
     return embedding.data[0].embedding
 
-# Search and retrieve data items that are relevant to the user query from the Pinecone index, to create a context
+
 @cl.step(name="Retrieve", type="retrieval")
-async def retrieve(embedding):
+async def retrieve(embedding, top_k):
+    """
+    Retrieve top_k closest vectors from the Pinecone index using the provided embedding.
+    """
     if pinecone_index == None:
         raise Exception("Pinecone index not initialized")
-    response = pinecone_index.query(vector=embedding, top_k=5, include_metadata=True)
-
-    return response
-
-# query the LLM with the query and the context
-@cl.step(name="LLM", type="llm")
-async def llm(
-    prompt,
-    chat_model="gpt-4-turbo-preview",
-):
-    messages = cl.user_session.get("messages", [])
-    messages.append(prompt)
-    settings = {"temperature": 0, "stream": True, "model": chat_model}
-    stream = await openai_client.chat.completions.create(messages=messages, **settings)
-    message = cl.message.Message(content="")
-    await message.send()
+    response = pinecone_index.query(
+        vector=embedding, top_k=top_k, include_metadata=True
+    )
+    return response.to_dict()
 
-    async for part in stream:
-        if token := part.choices[0].delta.content or "":
-            await message.stream_token(token)
 
-    await message.update()
-    messages.append({"role": "assistant", "content": message.content})
-    cl.user_session.set("messages", messages)
-    return message.content
+@cl.step(name="Retrieval", type="tool")
+async def get_relevant_documentation_chunks(question, top_k=5):
+    """
+    Retrieve relevant documentation chunks based on the question embedding.
+    """
+    embedding = await embed(question)
 
+    retrieved_chunks = await retrieve(embedding, top_k)
 
-@cl.step(name="Query", type="run")
-async def run(query):
-    embedding = await embed(query)
-    stored_embeddings = await retrieve(embedding)
-    contexts = []
-    prompt = literal_client.api.get_prompt(name="RAG prompt")
+    contexts = [match["metadata"]["text"] for match in retrieved_chunks["matches"]]
 
-    if not prompt:
-        raise Exception("Prompt not found")
-    for match in stored_embeddings["matches"]:
-        contexts.append(match["id"])
+    return "\n".join(contexts)
 
-    completion = await llm(prompt.format({"context": contexts, "question": query})[-1])
+async def run_multiple(tool_calls):
```

Comment:
I don't understand what run_multiple and run_single do. It seems complicated to me, the guide should be easy to follow. So it either needs a more explanation / comments, or simplification of the code.

Reply:
Agreed, I will remove this section from the guide.